### PR TITLE
Step3 - Similar movies screen 

### PR DESCRIPTION
**Pull request by sagilorber**
Similar movies screen added , you can see it by clicking on the similar movies link on the movie details page. this screen is a view pager that you can scroll horizontally through the  similar movies.
similar-movies-demo

### CODE REVIEW

File: .idea/vcs.xml [hunk 1]
```diff
@@ -2,5 +2,6 @@
 <project version="4">
   <component name="VcsDirectoryMappings">
     <mapping directory="$USER_HOME$" vcs="Git" />
+    <mapping directory="$PROJECT_DIR$" vcs="Git" />
```
Comment (gardenofwine) [inline .idea/vcs.xml]: I think you can add the whole `.idea` directory to the `.gitignore` file. That was you won't see these files (which are unrelated to the project) in the repository.

Comment (ifeins) [inline .idea/vcs.xml]: Actually, you shouldn't put the whole folder in the repository.
There are some project settings that it's useful to be committed.

File: .idea/misc.xml [hunk 2]
```diff
@@ -37,7 +37,7 @@
     <ConfirmationsSetting value="0" id="Add" />
     <ConfirmationsSetting value="0" id="Remove" />
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="true" assert-keyword="true" jdk-15="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_7" default="true" assert-keyword="true" jdk-15="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
Comment (ifeins) [inline .idea/misc.xml]: Just out of curiosity, why was it changed?
Also the `project-jdk-name` is still `1.8`

File: app/src/main/java/com/example/slorber/moviefiend/Movie.java [hunk 3]
```diff
@@ -48,6 +51,7 @@ public Movie createFromParcel(Parcel in) {
     public String getPosterPath(){return mPosterPath;}
     public String getOverview(){return mOverview;}
     public float getVotes(){return mVoteAverage;}
+    public int getId(){return mId;}
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/Movie.java]: format class

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java [hunk 4]
```diff
@@ -69,5 +70,10 @@ public void onClick(View v) {
 
     }
 
-    
+    public void onLabelClick(View v) {
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java]: Not really an indicative name.

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java [hunk 5]
```diff
@@ -69,5 +70,10 @@ public void onClick(View v) {
...
+        Intent intent = new Intent(this,SimilarMoviesActivity.class);
+        intent.putExtra(SimilarMoviesActivity.EXTRA_ID,movie.getId());
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java]: format class

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailsView.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * This view includes the movie poster, rating and overview.
+ * after inflated setItem is called with the movie object to assign the content.
+ * if called from similar movies screen the the "similar" link is hidden.
+ * Created by slorber on 12/09/2016.
+ */
+public class MovieDetailsView extends ScrollView {
+
+    private static String IMAGE_URL = "http://image.tmdb.org/t/p/w500/";
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailsView.java]: constants should be `final`

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailsView.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    public MovieDetailsView(Context context, AttributeSet attrs, int defStyle) {
+        super(context, attrs, defStyle);
+        LayoutInflater.from(context).inflate(R.layout.movie_details_child, this, true);
+        setupChildren();
+    }
+
+    private void setupChildren() {
+
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailsView.java]: format class

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailsView.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        mDescriptionTextView = (TextView) findViewById(R.id.overview);
+        mImageView = (ImageView) findViewById(R.id.movie_large_image);
+        mStar = (RatingView)findViewById(R.id.movie_star);
+    }
+
+    public void removeView(){
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailsView.java]: Not a good name for a method.
Which view is being removed? (it's actually not removed, but hidden)

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailsView.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        ((TextView)findViewById(R.id.similar_movies_label)).setVisibility(View.GONE);
+    }
+
+    public void setItem(Movie movie) {
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailsView.java]: `setMovie`

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+import org.json.JSONObject;
+
+import java.util.List;
+import java.util.Locale;
+
+import me.relex.circleindicator.CircleIndicator;
+
+public class SimilarMoviesActivity extends AppCompatActivity implements LoaderManager.LoaderCallbacks<List<Movie>>{
+
+    public static String EXTRA_ID = "id";
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: `final`

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    private CircleIndicator mIndicator;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_similar_movies);
+        mPager = (ViewPager) findViewById(R.id.view_pager);
+        mIndicator = (CircleIndicator) findViewById(R.id.indicator);
+        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
+        setSupportActionBar(toolbar);
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: You can inline the `toolbar` variable, if you're only using it once.

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        return new GetMoviesLoader(this,String.format(Locale.getDefault(),URL, getIntent().getIntExtra(EXTRA_ID,0)));
+    }
+
+    @Override
+    public void onLoadFinished(Loader<List<Movie>> loader, List<Movie> data) {
+        if(data == null){
+            Toast.makeText(this,"Server error... :(",Toast.LENGTH_LONG).show();
+        }
+        else if(data.size()==0)
+        {
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: format

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
+        setSupportActionBar(toolbar);
+        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+        getSupportLoaderManager().initLoader(1, null, this);
+
+    }
+
+    @Override
+    public Loader<List<Movie>> onCreateLoader(int id, Bundle args) {
+        return new GetMoviesLoader(this,String.format(Locale.getDefault(),URL, getIntent().getIntExtra(EXTRA_ID,0)));
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: Don't use string concatenation to build URLs.
Use `Uri.Builder`

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        else if(data.size()==0)
+        {
+            ((ViewFlipper)findViewById(R.id.view_flipper)).showNext();
+        }
+        else {
+            mPagerAdapter = new ScreenSlidePagerAdapter(this, data);
+            mPager.setAdapter(mPagerAdapter);
+            mIndicator.setViewPager(mPager);
+        }
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: You can remove nesting from this class by "exiting" on edge conditions.
For example:

```
if (data == null) {
  Toast...
  return;
}

if (data.size() == 0) {
  showNext;
  return
}

// main bulk of method that handles the common case without any nesting.
```

This really improves readability.

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Override
+    public void onLoadFinished(Loader<List<Movie>> loader, List<Movie> data) {
+        if(data == null){
+            Toast.makeText(this,"Server error... :(",Toast.LENGTH_LONG).show();
+        }
+        else if(data.size()==0)
+        {
+            ((ViewFlipper)findViewById(R.id.view_flipper)).showNext();
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: What exactly does `showNext` does?

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+        }
+
+        @Override
+        public int getCount() {
+            return mMovieList.size();
+        }
+
+        @Override
+        public boolean isViewFromObject(View arg0, Object arg1) {
+            return arg0 == ((View) arg1);
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: No need for casting if you're comparing references.

File: app/src/main/res/layout/activity_movie_detail.xml [hunk 17]
```diff
@@ -49,47 +49,18 @@
...
-                app:drawable_margin="4dp"
-                app:drawable_size="14dp"
-                app:is_indicator="false"
-                app:max_count="5"
-                app:rating="3.5"
-                android:gravity="center"
-                android:orientation="vertical"
-                android:padding="16dp"/>
-        <TextView
+        <com.example.slorber.moviefiend.MovieDetailsView
```
Comment (ifeins) [inline app/src/main/res/layout/activity_movie_detail.xml]: 👍

File: app/src/main/res/layout/activity_similar_movies.xml [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+
+        <TextView
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:textAppearance="?android:attr/textAppearanceMedium"
+            android:text="No similar movies....:("
+            android:id="@+id/textView"
+            android:layout_marginTop="?attr/actionBarSize"
+            android:gravity="center"/>
```
Comment (ifeins) [inline app/src/main/res/layout/activity_similar_movies.xml]: I'm guessing that's what will be displayed when invoking `showNext`?

File: app/src/main/res/layout/movie_details_child.xml [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                app:drawable_empty="@drawable/ic_star_empty"
+                app:drawable_filled="@drawable/ic_star_filled"
+                app:drawable_half="@drawable/ic_star_half"
+                app:drawable_margin="4dp"
+                app:drawable_size="14dp"
+                app:is_indicator="false"
+                app:max_count="5"
+                app:rating="3.5"
```
Comment (ifeins) [inline app/src/main/res/layout/movie_details_child.xml]: Are you using this for being able to preview the view in the Android Studio designer?
If so, that's awesome, but is it possible to use the `tools` namespace instead?

File: app/src/main/res/layout/movie_details_child.xml [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+                android:padding="@dimen/activity_horizontal_margin"
+                />
+            <TextView
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:padding="8dp"
+                android:gravity="center"
+                android:id="@+id/similar_movies_label"
+                android:onClick="onLabelClick"
+                android:clickable="true"
```
Comment (ifeins) [inline app/src/main/res/layout/movie_details_child.xml]: interesting :)

File: app/src/main/res/layout/activity_similar_movies.xml (hunk 18 above)
Comment (sagilorber) [inline app/src/main/res/layout/activity_similar_movies.xml]: Yep..

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java (hunk 15 above)
Comment (sagilorber) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: swap the view pager with a textview showing no similar movies.

File: app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java (hunk 13 above)
Comment (sagilorber) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: but the loader gets it as a String and he convert it to Uri..

Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/SimilarMoviesActivity.java]: So you can change the loader to receive a `Uri`

**Pull request merged**